ceph: enhancement disk sanitize options (#5452)
rook/rook: Documentation/ceph-cluster-crd.md

  cephVersion:
    image: ceph/ceph:v14.2.9 # Should match external cluster version
```

### Cleanup policy

Rook has the ability to cleanup resources and data that were deployed.
The policy represents the confirmation that cluster data should be forcibly deleted.
The cleanupPolicy should only be added to the cluster when the cluster is about to be deleted.

> After any field of the cleanup policy is set, Rook will stop configuring the cluster as if the cluster is about to be destroyed in order to prevent these settings from being deployed unintentionally.

Author: True. Actually it's not my write-up I just moved it but will fix.

The `cleanupPolicy` CR settings has different fields:

* `confirmation`: If `yes-really-destroy-data` the operator will automatically delete data on the hostpath of cluster nodes and clean devices with OSDs when a `delete cephcluster` command is issued. Only `yes-really-destroy-data` and an empty string are valid values for this field.

> * sanitizeDisks: sanitizeDisks represents advanced settings that can be used to sanitize drives.

Reviewer: ```suggestion
* diskPolicy: advanced settings that can be used to sanitize drives.
```

By default Rook, will only remove Ceph's metadata, which allows a cluster re-installation if desired.
However, the administrator might want to sanitize the drives in more depth, that's what the following flags do:

>   * method: indicates if the entire disk should be sanitized or simply ceph's metadata. Possible choices are 'data' or 'metadata' (default)

Reviewer: "data" and "metadata" could be confused with the device set options. What about this?
```suggestion
  * method: indicates if the entire disk should be sanitized or simply ceph's metadata. Possible choices are 'complete' or 'quick' (default)
```

>   * utility: indicates the tool to use to sanitize the disk. Possible choices are 'dd' (default) or 'shred'.

Reviewer: Why do we need to expose the command line tool in the CR? This is an implementation detail of the policy. Can't we figure out internally which tool to use based on the settings?

Author: it's probably doable, but exposing the `utility` leaves the door open to more utilities. That's the intent.

Reviewer: I still don't see the need to expose the command line tool. If another utility is needed, it would be driven by some other settings in the CR.

Author: Fixed!

  The shred utility has more elaborate options and is more powerful
  * dataSource: indicate where to get random bytes from to write on the disk. Possible choices are 'zero' (default), 'random' or 'urandom'.
  Using random sources will consume entropy from the system and will take much more time then the zero source

>   * iteration: overwrite N times instead of the default (1). Takes an integer value, only supported when 'shred' utility is selected

Reviewer: If iteration > 1 we can automatically use shred, right?

Author: Yes we could

> To activate the cleanup, you can use the following command **AT YOUR OWN RISK**:

Reviewer: ```suggestion
To automated activation of the cleanup, you can use the following command **AT YOUR OWN RISK**:
```

Author: I think you meant `s/automated/automate/`, fixing.

```console
kubectl -n rook-ceph patch cephcluster rook-ceph --type merge -p '{"spec":{"cleanupPolicy":{"confirmation":"yes-really-destroy-data"}}}'
```

When applied you will see the following from the Operator's logs:

```text
2020-05-27 13:24:04.267665 I | ceph-spec: CR has changed for "rook-ceph". diff=  v1.ClusterSpec{
        ... // 16 identical fields
        Mgr:                            v1.MgrSpec{},
        RemoveOSDsIfOutAndSafeToRemove: false,
-       CleanupPolicy:                  v1.CleanupPolicySpec{},
+       CleanupPolicy:                  v1.CleanupPolicySpec{Confirmation: "yes-really-destroy-data"},
  }
```

Nothing will happen until the deletion of the CR is requested, so this can still be reverted.

> However, all new orchestration/reconciliation will be blocked with this cleanup policy enabled.

Reviewer: ```suggestion
However, all new configuration by the operator will be blocked with this cleanup policy enabled.
```

The policy represents the confirmation that cluster data should be forcibly deleted.
The cleanupPolicy should only be added to the cluster when the cluster is about to be deleted.
After the `confirmation` field of the cleanup policy is set, Rook will stop configuring the cluster as if the cluster is about to be destroyed in order to prevent these settings from being deployed unintentionally.
The `cleanupPolicy` CR settings has different fields:

* `confirmation`: If `yes-really-destroy-data` the operator will automatically delete data on the hostpath of cluster nodes and clean devices with OSDs when a `delete cephcluster` command is issued. Only `yes-really-destroy-data` and an empty string are valid values for this field.
* sanitizeDisks: sanitizeDisks represents advanced settings that can be used to sanitize drives.
By default Rook, will only remove Ceph's metadata, which allows a cluster re-installation if desired.
However, the administrator might want to sanitize the drives in more depth with the following flags:
  * method: indicates if the entire disk should be sanitized or simply ceph's metadata. Possible choices are 'complete' or 'quick' (default)

>   * dataSource: indicate where to get random bytes from to write on the disk. Possible choices are 'zero' (default), 'random' or 'urandom'.

Reviewer 2: I hesitate to accept 'random' since it would block.

Since there is not so big difference of randomness between  /dev/random and /dev/urandom, I suggest removing 'random' or mentioning 'random' may block.

Myths about /dev/urandom:
https://www.2uo.de/myths-about-urandom/

Author: Agreed, I'd even let `shred` pick the right random instead of passing one if the source is not zero.

  cephVersion:
    image: ceph/ceph:v15.2.4 # Should match external cluster version
```

### Cleanup policy

Rook has the ability to cleanup resources and data that were deployed.
The policy represents the confirmation that cluster data should be forcibly deleted.
The cleanupPolicy should only be added to the cluster when the cluster is about to be deleted.
After the `confirmation` field of the cleanup policy is set, Rook will stop configuring the cluster as if the cluster is about to be destroyed in order to prevent these settings from being deployed unintentionally.
The `cleanupPolicy` CR settings has different fields:

* `confirmation`: If `yes-really-destroy-data` the operator will automatically delete data on the hostpath of cluster nodes and clean devices with OSDs when a `delete cephcluster` command is issued. Only `yes-really-destroy-data` and an empty string are valid values for this field.
* `sanitizeDisks`: sanitizeDisks represents advanced settings that can be used to sanitize drives.

> By default Rook, will only remove Ceph's metadata, which allows a cluster re-installation if desired.

Reviewer: ```suggestion
By default if the `confirmation` is set, Rook will only remove Ceph's metadata, which allows a cluster re-installation if desired.
```

However, the administrator might want to sanitize the drives in more depth with the following flags:

>   * `method`: indicates if the entire disk should be sanitized or simply ceph's metadata. Possible choices are 'complete' or 'quick' (default)

Reviewer: ```suggestion
  * `method`: indicates if the entire disk should be sanitized or simply ceph's metadata. Possible choices are 'quick' (default) or 'complete'
```

  * `dataSource`: indicate where to get random bytes from to write on the disk. Possible choices are 'zero' (default) or 'random'.
  Using random sources will consume entropy from the system and will take much more time then the zero source
  * `iteration`: overwrite N times instead of the default (1). Takes an integer value

> To automate activation of the cleanup, you can use the following command **AT YOUR OWN RISK**:

Reviewer: ```suggestion
To automate activation of the cleanup, you can use the following command. **WARNING: DATA WILL BE PERMANENTLY DELETED**:
```

```console
kubectl -n rook-ceph patch cephcluster rook-ceph --type merge -p '{"spec":{"cleanupPolicy":{"confirmation":"yes-really-destroy-data"}}}'
```

> When applied you will see the following from the Operator's logs:

Reviewer: I don't think we need to show the detailed output from the log. I'd suggest removing lines 1027-1037

Author: I thought it was a useful indication but it might change so let's remove it. We already have a message about the cleanup policy being detected.
